File: src/main/java/com/tambapps/fft4j/fourier/filtering/Filter.java
package com.tambapps.fft4j.fourier.filtering;

import com.tambapps.fft4j.complex.array2d.CArray2D;

/**
 * A filter of a Fourier Transform
 */
public interface Filter {

  /**
   * Applies the filter on the given array
   *
   * @param array the array to filter
   */
  void apply(CArray2D array);

  /**
   * same as apply (util for Groovy operator)
   *
   * @param array the array to filter
   */
  void call(CArray2D array);

  /**
   * Applies the filter on a copied array and returns it
   *
   * @param array the array to filter
   * @return the filtered array
   */
  CArray2D applied(CArray2D array);

}

File: src/main/kotlin/com/tambapps/fft4j/fourier/filtering/Filter.kt
package com.tambapps.fft4j.fourier.filtering

import com.tambapps.fft4j.complex.array2d.CArray2D

/**
 * A filter of a Fourier Transform
 */
interface Filter {
    /**
     * Applies the filter on the given array
     *
     * @param array the array to filter
     */
    fun apply(array: CArray2D)

    /**
     * same as apply
     *
     * @param array the array to filter
     */
    operator fun invoke(array: CArray2D)

    /**
     * Applies the filter on a copied array and returns it
     *
     * @param array the array to filter
     * @return the filtered array
     */
    fun applied(array: CArray2D): CArray2D
}
